File: backend/app/models/payroll.py
from datetime import date, datetime, time
from decimal import Decimal
from uuid import UUID

from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Integer,
    JSON,
    Numeric,
    String,
    Text,
    Time,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database.base import Base
from app.models.mixins import TimestampMixin, UUIDPrimaryKeyMixin


class PayrollRun(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "payroll_runs"
    __table_args__ = (
        UniqueConstraint("company_id", "month", name="uq_payroll_runs_company_month"),
    )

    company_id: Mapped[UUID] = mapped_column(ForeignKey("companies.id"), nullable=False, index=True)
    month: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="draft", index=True)
    created_by_admin_user_id: Mapped[UUID] = mapped_column(
        ForeignKey("admin_users.id"), nullable=False
    )
    approved_by_admin_user_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("admin_users.id"), nullable=True
    )
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    locked_by_admin_user_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("admin_users.id"), nullable=True
    )
    locked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    paid_by_admin_user_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("admin_users.id"), nullable=True
    )
    paid_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    entries = relationship("PayrollEntry", back_populates="run", cascade="all, delete-orphan")


class PayrollEntry(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "payroll_entries"
    __table_args__ = (
        UniqueConstraint("payroll_run_id", "employee_id", name="uq_payroll_entries_run_employee"),
    )

    company_id: Mapped[UUID] = mapped_column(ForeignKey("companies.id"), nullable=False, index=True)
    payroll_run_id: Mapped[UUID] = mapped_column(
        ForeignKey("payroll_runs.id", ondelete="CASCADE"), nullable=False, index=True
    )
    employee_id: Mapped[UUID] = mapped_column(
        ForeignKey("employees.id"), nullable=False, index=True
    )
    team_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    job_title: Mapped[str | None] = mapped_column(String(255), nullable=True)
    salary_type: Mapped[str] = mapped_column(String(20), nullable=False, default="monthly")
    currency: Mapped[str] = mapped_column(String(3), nullable=False, default="EGP")
    salary_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False, default=0)
    hourly_rate: Mapped[Decimal] = mapped_column(Numeric(14, 4), nullable=False, default=0)
    expected_work_days: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    expected_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    worked_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    approved_manual_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    pending_manual_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    rejected_manual_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    idle_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    late_minutes: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    paid_break_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    unpaid_break_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    absence_days: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    recorded_overtime_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    approved_overtime_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    rejected_overtime_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    overtime_eligible: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)

    deduct_lateness: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    lateness_deduction_amount: Mapped[Decimal] = mapped_column(
        Numeric(14, 2), nullable=False, default=0
    )
    lateness_note: Mapped[str | None] = mapped_column(Text, nullable=True)
    deduct_idle: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    idle_deduction_amount: Mapped[Decimal] = mapped_column(
        Numeric(14, 2), nullable=False, default=0
    )
    idle_note: Mapped[str | None] = mapped_column(Text, nullable=True)
    pay_overtime: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    overtime_decision: Mapped[str] = mapped_column(String(20), nullable=False, default="pending")
    overtime_multiplier: Mapped[Decimal] = mapped_column(Numeric(8, 2), nullable=False, default=1)
    custom_overtime_amount: Mapped[Decimal | None] = mapped_column(Numeric(14, 2), nullable=True)
    overtime_note: Mapped[str | None] = mapped_column(Text, nullable=True)
    deduct_unpaid_breaks: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    unpaid_break_deduction_amount: Mapped[Decimal] = mapped_column(
        Numeric(14, 2), nullable=False, default=0
    )
    unpaid_break_note: Mapped[str | None] = mapped_column(Text, nullable=True)
    bonus_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False, default=0)
    additional_deduction_amount: Mapped[Decimal] = mapped_column(
        Numeric(14, 2), nullable=False, default=0
    )
    adjustment_note: Mapped[str | None] = mapped_column(Text, nullable=True)
    base_salary: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False, default=0)
    overtime_amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False, default=0)
    total_deductions: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False, default=0)
    total_bonuses: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False, default=0)
    final_salary: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False, default=0)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="draft", index=True)
    calculation_snapshot: Mapped[dict | None] = mapped_column(JSON, nullable=True)

    run = relationship("PayrollRun", back_populates="entries")
    employee = relationship("Employee")
    adjustments = relationship(
        "PayrollAdjustment", back_populates="entry", cascade="all, delete-orphan"
    )


class PayrollAdjustment(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "payroll_adjustments"

    company_id: Mapped[UUID] = mapped_column(ForeignKey("companies.id"), nullable=False, index=True)
    payroll_entry_id: Mapped[UUID] = mapped_column(
        ForeignKey("payroll_entries.id", ondelete="CASCADE"), nullable=False, index=True
    )
    adjustment_type: Mapped[str] = mapped_column(String(40), nullable=False, index=True)
    amount: Mapped[Decimal] = mapped_column(Numeric(14, 2), nullable=False)
    reason: Mapped[str] = mapped_column(Text, nullable=False)
    created_by_admin_user_id: Mapped[UUID] = mapped_column(
        ForeignKey("admin_users.id"), nullable=False
    )

    entry = relationship("PayrollEntry", back_populates="adjustments")


class WorkScheduleOverride(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "work_schedule_overrides"

    company_id: Mapped[UUID] = mapped_column(ForeignKey("companies.id"), nullable=False, index=True)
    employee_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("employees.id"), nullable=True, index=True
    )
    scope: Mapped[str] = mapped_column(String(20), nullable=False)
    override_type: Mapped[str] = mapped_column(String(20), nullable=False, index=True)
    effective_date: Mapped[date | None] = mapped_column(Date, nullable=True, index=True)
    permanent: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    shift_start: Mapped[time | None] = mapped_column(Time(), nullable=True)
    shift_end: Mapped[time | None] = mapped_column(Time(), nullable=True)
    break_rules: Mapped[list[dict] | None] = mapped_column(JSON, nullable=True)
    reason: Mapped[str] = mapped_column(Text, nullable=False)
    created_by_admin_user_id: Mapped[UUID] = mapped_column(
        ForeignKey("admin_users.id"), nullable=False
    )

File: frontend/admin-dashboard/src/api/payroll.ts
import { apiFetch, apiFile, withQuery } from "@/api/client";

export type PayrollStatus = "draft" | "needs_review" | "approved" | "locked" | "paid";

export type PayrollEntry = {
  id: string;
  employee_id: string;
  employee_name: string;
  team?: string | null;
  job_title?: string | null;
  salary_type: "monthly" | "hourly";
  salary: number;
  currency: string;
  hourly_rate: number;
  expected_work_days: number;
  worked_days: number;
  leave_days: number;
  expected_seconds: number;
  worked_seconds: number;
  normal_seconds: number;
  total_payable_seconds: number;
  approved_manual_seconds: number;
  pending_manual_seconds: number;
  rejected_manual_seconds: number;
  idle_seconds: number;
  late_minutes: number;
  raw_late_minutes: number;
  early_leave_minutes: number;
  paid_break_seconds: number;
  unpaid_break_seconds: number;
  absence_days: number;
  recorded_overtime_seconds: number;
  approved_overtime_seconds: number;
  rejected_overtime_seconds: number;
  overtime_eligible: boolean;
  deduct_lateness: boolean;
  lateness_deduction_amount: number;
  lateness_note?: string | null;
  deduct_idle: boolean;
  idle_deduction_amount: number;
  idle_note?: string | null;
  deduct_unpaid_breaks: boolean;
  unpaid_break_deduction_amount: number;
  unpaid_break_note?: string | null;
  pay_overtime: boolean;
  overtime_decision: "pending" | "paid" | "rejected";
  overtime_multiplier: number;
  custom_overtime_amount?: number | null;
  overtime_note?: string | null;
  bonus_amount: number;
  additional_deduction_amount: number;
  adjustment_note?: string | null;
  base_salary: number;
  overtime_amount: number;
  total_deductions: number;
  total_bonuses: number;
  final_salary: number;
  status: PayrollStatus;
  calculation?: Record<string, number | string | boolean | null>;
  adjustments?: PayrollAdjustment[];
};

export type PayrollAdjustment = {
  id: string;
  type: string;
  amount: number;
  reason: string;
  created_by: string;
  created_at: string;
};

export type PayrollSheet = {
  run: {
    id: string;
    month: string;
    status: PayrollStatus;
    approved_at?: string | null;
    locked_at?: string | null;
    paid_at?: string | null;
    period_start?: string | null;
    period_end?: string | null;
    cycle_timezone?: string | null;
  };
  settings: PayrollSettings;
  summary: {
    employees: number;
    needs_review: number;
    late_employees: number;
    overtime_employees: number;
    currencies: Record<
      string,
      { base: number; overtime: number; bonuses: number; deductions: number; final: number }
    >;
  };
  teams: string[];
  entries: PayrollEntry[];
};

export type PayrollFilters = {
  month: string;
  start_date?: string;
  end_date?: string;
  team?: string;
  employee_id?: string;
  status?: string;
  overtime_eligible?: boolean;
  has_lateness?: boolean;
  has_idle?: boolean;
  has_deductions?: boolean;
  has_manual_adjustments?: boolean;
};

export type PayrollEntryUpdate = Partial<
  Pick<
    PayrollEntry,
    | "deduct_lateness"
    | "lateness_deduction_amount"
    | "lateness_note"
    | "deduct_idle"
    | "idle_deduction_amount"
    | "idle_note"
    | "deduct_unpaid_breaks"
    | "unpaid_break_deduction_amount"
    | "unpaid_break_note"
    | "overtime_decision"
    | "overtime_multiplier"
    | "custom_overtime_amount"
    | "overtime_note"
    | "bonus_amount"
    | "additional_deduction_amount"
    | "adjustment_note"
    | "status"
  >
>;

export function getPayrollSheet(filters: PayrollFilters) {
  return apiFetch<PayrollSheet>(withQuery("/payroll/sheet", filters));
}

export function getPayrollEntry(entryId: string) {
  return apiFetch<PayrollEntry>(`/payroll/entries/${entryId}`);
}

export function updatePayrollEntry(entryId: string, input: PayrollEntryUpdate) {
  return apiFetch<PayrollEntry>(`/payroll/entries/${entryId}`, {
    method: "PATCH",
    body: JSON.stringify(input),
  });
}

export function addPayrollAdjustment(
  entryId: string,
  input: { adjustment_type: string; amount: number; reason: string },
) {
  return apiFetch<PayrollEntry>(`/payroll/entries/${entryId}/adjustments`, {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export function removePayrollAdjustment(adjustmentId: string) {
  return apiFetch<{ deleted: boolean }>(`/payroll/adjustments/${adjustmentId}`, {
    method: "DELETE",
  });
}

export function updatePayrollRunStatus(
  runId: string,
  status: "draft" | "approved" | "locked" | "paid",
  reason?: string,
) {
  return apiFetch<PayrollSheet["run"]>(`/payroll/runs/${runId}/status`, {
    method: "PATCH",
    body: JSON.stringify({ status, reason }),
  });
}

export function getPayrollExceptions(
  month: string,
  range?: { start_date?: string; end_date?: string },
) {
  return apiFetch<Record<string, PayrollEntry[]>>(
    withQuery("/payroll/exceptions", { month, ...range }),
  );
}

export type PayrollSettings = {
  cycle_start_day: number;
  cycle_end_day: number;
  timezone: string;
};

export function getPayrollSettings() {
  return apiFetch<PayrollSettings>("/payroll/settings");
}

export function updatePayrollSettings(input: PayrollSettings) {
  return apiFetch<PayrollSettings>("/payroll/settings", {
    method: "PATCH",
    body: JSON.stringify(input),
  });
}

export function createScheduleOverride(input: {
  scope: "employee" | "employees" | "team" | "company";
  override_type: "shift" | "breaks" | "both";
  employee_id?: string;
  employee_ids?: string[];
  team_id?: string;
  effective_date?: string;
  permanent: boolean;
  shift_start?: string;
  shift_end?: string;
  break_rules?: Array<{
    name: string;
    minutes: number;
    paid: boolean;
    start_time: string;
    end_time: string;
  }>;
  reason: string;
}) {
  return apiFetch<{ id: string; affected_employees: number }>("/payroll/schedule-overrides", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export type ScheduleOverride = {
  id: string;
  scope: "employee" | "team" | "company";
  override_type: "shift" | "breaks" | "both";
  employee_id?: string | null;
  employee_name?: string | null;
  team_id?: string | null;
  team_name?: string | null;
  effective_date?: string | null;
  break_rules?: Array<{
    name: string;
    minutes: number;
    paid: boolean;
    start_time?: string | null;
    end_time?: string | null;
  }> | null;
  shift_start?: string | null;
  shift_end?: string | null;
  reason: string;
  created_at: string;
};

export function listScheduleOverrides(upcomingOnly = true) {
  return apiFetch<ScheduleOverride[]>(
    withQuery("/payroll/schedule-overrides", { upcoming_only: upcomingOnly }),
  );
}

export function deleteScheduleOverride(overrideId: string) {
  return apiFetch<{ deleted: boolean }>(`/payroll/schedule-overrides/${overrideId}`, {
    method: "DELETE",
  });
}

export async function downloadPayroll(filters: PayrollFilters, format: "csv" | "excel" | "pdf") {
  const blob = await apiFile(withQuery("/payroll/export", { ...filters, format }));
  const extension = format === "excel" ? "xlsx" : format;
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = `payroll-${filters.month}.${extension}`;
  document.body.appendChild(anchor);
  anchor.click();
  anchor.remove();
  URL.revokeObjectURL(url);
}
